File: submission/12017800/code/07_decompose_random_coop/BestIndividual.java
import java.util.concurrent.locks.ReentrantLock;
import java.util.ArrayList;
import java.util.List;

public class BestIndividual extends Individual {
	private Individual backup;
	private double fitness;
	private OptimizationFunction function;
	private final ReentrantLock mutex = new ReentrantLock();

	public BestIndividual(int numCells, OptimizationFunction function) {
		super(numCells, function.getMin(), function.getMax());
		this.backup = super.clone();
		this.function = function;
		this.fitness = function.getFitness(this);
	}

	public BestIndividual(BestIndividual other) {
		super((Individual)other);
		this.backup = super.clone();
		this.function = other.function;
		this.fitness = other.fitness;
	}

	public double compareAndSet(Individual individual, List<Integer> indices) {
		mutex.lock();
		Individual tmpIndividual = new Individual(this);

		for (int i = 0; i < indices.size(); i++) {
			int index = indices.get(i);
			tmpIndividual.getCell(index).setValue(individual.getCell(i).getValue());
		}

		double tmpFitness = function.getFitness(tmpIndividual);

		if (tmpFitness < this.fitness) {

			for (int i = 0; i < indices.size(); i++) {
				int index = indices.get(i);
				super.getCell(index).setValue(individual.getCell(i).getValue());
			}

			tmpFitness = function.getFitness(tmpIndividual);

			if (tmpFitness < this.fitness) {
				fitness = tmpFitness;
				backup = super.clone();
			}
			else {
				restore(backup);
			}

		}

		mutex.unlock();
		return tmpFitness;
	}

	public void restore(Individual individual) {
		cells = new ArrayList<>();

		for (int i = 0; i < individual.getNumCells(); i++) {
			cells.add(individual.getCell(i).clone());
		}
	}
}

File: results/02 bak/bestIndividual.cpp
#include "bestIndividual.h"

BestIndividual::BestIndividual(size_t numCells, OptimizationFunction* function) { //function->getMin()
	this->numCells = numCells;
	this->min = function->getMin();
	this->max = function->getMax();
	this->function = function;

	for (int i = 0; i < numCells; i++) {
		Cell cell(min, max);
		cells.push_back(cell);
	}

	backUp = clone();
}

BestIndividual::~BestIndividual() {

}

long double BestIndividual::compareAndSet(Individual* individual, int startCellIndex, int endCellIndex) {
	Individual newIndividual(*this);
	// cout << "BEFORE: " << toString() << endl;
	for (int i = 0; i < endCellIndex - startCellIndex; i++) {
		newIndividual[startCellIndex + i] = (*individual)[i];
	}

	long double individualFitness = function->getFitness(&newIndividual);
	// cout << "DURING: " << toString() << " FITNESS: " << individualFitness << " VS " << fitness << endl;

	if (individualFitness < fitness) {
		lock.lock();

		for (int i = 0; i < endCellIndex - startCellIndex; i++) {
			(*this)[startCellIndex + i] = (*individual)[i];
		}

		individualFitness = function->getFitness(&newIndividual);
		
		if (individualFitness < fitness) {
			fitness = individualFitness;
			backUp = clone();
		}
		else {
			deepCopy(&backUp);
		}

		lock.unlock();
	}
	// cout << "AFTER: " << toString() << endl;



	return individualFitness;
}
